common-api/src/main/scala/pl/touk/nussknacker/engine/graph/expression/TabularTypedData.scala

```diff
@@ -64,32 +80,40 @@ object TabularTypedData {
     }
   }
 
-  private def validateColumnNamesUniqueness(
-      columns: Vector[Column.Definition]
-  ): Either[HumanReadableErrorMessage, Unit] = {
+  private def validateColumnNamesUniqueness(columns: Vector[Column.Definition]): Either[CreationError, Unit] = {
     val originColumnNames = columns.map(_.name)
     val duplicates        = originColumnNames.diff(originColumnNames.distinct)
-    if (duplicates.isEmpty) Right(())
-    else Left(s"Column names should be unique. Duplicates: ${duplicates.distinct.mkString(",")}")
+    NonEmptyList.fromFoldable(duplicates) match {
+      case None      => Right(())
+      case Some(nel) => Left(ColumnNameUniquenessViolation(nel))
+    }
   }
 
   private def validateCellsCount(columns: Vector[Column.Definition], rows: Vector[Vector[Any]]) = {
-    rows.foldLeft(Right(()): Either[HumanReadableErrorMessage, Unit]) {
+    rows.foldLeft(Right(()): Either[CreationError, Unit]) {
       case (r @ Right(()), row) if row.size == columns.size => r
-      case (Right(()), _)   => Left("All rows should have the same number of cells as there are columns")
-      case (l @ Left(_), _) => l
+      case (Right(()), _)                                   => Left(CellsCountInRowDifferentThanColumnsCount)
+      case (l @ Left(_), _)                                 => l
     }
   }
 
-  private def validateCellValuesType(data: TabularTypedData): Either[HumanReadableErrorMessage, Unit] = {
-    data.rows.flatMap(_.cells).foldLeft(Right(()): Either[HumanReadableErrorMessage, Unit]) {
-      case (r @ Right(()), cell) if cell.rawValue == RawValue(null) => r
-      case (r @ Right(()), cell) if doesCellValueLookOK(cell)       => r
-      case (Right(()), cell) =>
-        Left(
-          s"Column has a '${cell.definition.aType.getCanonicalName}' type but the value '${cell.rawValue.value}' cannot be converted to it."
-        )
-      case (l @ Left(_), _) => l
+  private def validateCellValuesType(data: TabularTypedData): Either[CreationError, Unit] = {
+    val cellErrors = data.rows
+      .foldLeft(List.empty[CellCoordinates]) { case (acc, row) =>
```

Comment:
You don't change the previous accumulator so it can be replaced by `data.rows.flatMap(validateCellsInRow)`